pyteal/stack_frame.py

```diff
@@ -0,0 +1,739 @@
+        frame_infos = frame_infos[last_keep_idx : last_keep_idx + 1]
+
+        # 7. we finish by constructing a list[StackFrame] from our one remaining frame_info
+        self._frames = _make_stack_frames(frame_infos)
+
+    def __len__(self) -> int:
+        return len(self._frames)
+
+    def best(self) -> StackFrame:
+        """
+        Return the best guess as to the user-authored birthplace of the
+        associated StackFrame's
+        """
+        assert (
+            self._frames
+        ), f"expected to have some frames but currently {self._frames=}"
+        return self._frames[-1]
+
+    def __repr__(self) -> str:
+        return f"{'C' if self._compiler_gen else 'U'}{self._frames}"
+
+    def nodes(self) -> list[AST | None]:
+        return [f.node for f in self._frames]
+
+    @classmethod
+    def _walk_asts(cls, func: Callable[["Expr"], None], *exprs: "Expr") -> None:  # type: ignore
+        from pyteal.ast import (
+            Assert,
+            BinaryExpr,
+            Cond,
+            Expr,
+            Return,
+            Seq,
+            SubroutineDeclaration,
+        )
+        from pyteal.ast.frame import Proto
+
+        for expr in exprs:
+            e = cast(Expr, expr)
+            func(e)
+
+            match e:
+                case Assert():
+                    cls._walk_asts(func, *e.cond)
+                case BinaryExpr():
+                    cls._walk_asts(func, e.argLeft, e.argRight)
+                case Cond():
+                    cls._walk_asts(func, *(y for x in e.args for y in x))
+                case Proto():
+                    cls._walk_asts(func, e.mem_layout)
+                case Seq():
+                    cls._walk_asts(func, *e.args)
+                case Return():
+                    cls._walk_asts(func, e.value)
+                case SubroutineDeclaration():
+                    cls._walk_asts(func, e.body)
+                case _:
+                    # TODO: implement more cases, but no need to error as this isn't used for functionality's sake.
+                    pass
+
+    @classmethod
+    def _debug_asts(cls, *exprs: "Expr") -> None:  # type: ignore
+        """
+        For deubgging purposes only!
+        """
+        from pyteal.ast import Expr
+
+        if cls.sourcemapping_is_off():
+            return
+
+        def dbg(e: Expr):
+            print(
+                type(e), ": ", e.stack_frames.best().as_pyteal_frame().hybrid_unparsed()
+            )
+
+        cls._walk_asts(dbg, *exprs)
+
+    @classmethod
+    def mark_asts_as_compiler_gen(cls, *exprs: "Expr") -> None:  # type: ignore
+        from pyteal.ast import Expr
+
+        if cls.sourcemapping_is_off():
+            return
+
+        def mark(e: Expr):
+            e.stack_frames._compiler_gen = True
+
+        cls._walk_asts(mark, *exprs)
+
+    @classmethod
+    def reframe_asts(cls, stack_frames: "NatalStackFrame", *exprs: "Expr") -> None:  # type: ignore
```

Comment:
is it marking/overwriting certain ASTs with another `NatalStackFrame`?

Reply:
Yes, it is.